### PR TITLE
Move tzname from JS library code into C

### PR DESCRIPTION
**Pull request by kripken**
The main goal here is to work around current lld limitations, namely that we can't import a global location from JS there. See https://github.com/WebAssembly/tool-conventions/issues/48

This is a first step to a better solution, which would be to import the global location from compiled code, instead of importing a function we call to get that location.

There are some downsides to this PR,

 * It doesn't fix the lld issue, just works around it. As discussed in that issue, even if we remove all such usages from the emscripten system libraries, emscripten users may have their own. But that is likely very rare.
 * It adds overhead, now we call into compiled code to get the address. I doubt for these methods it matters, though, and as mentioned above we should eventually replace the function call with a global address once JS can import such things from wasm.
 * This uses the `deps_info.json` mechanism which is not the prettiest - it tells the linker that if we see a symbol, we need to link in more C code as JS code will need it to be there (otherwise, we'd get to the JS linking at the end and only then figure it out). But not a big deal either, I'd say. And we can hopefully improve that in time (autogenerate it and/or replace it).

cc @sbc100

### CODE REVIEW

**Comment by NWilson** [comment]
I'm still a bit surprised that this is needed - after the storage for the int is defined in C, I'd expect LLD's exported "fake globals" to be usable for getting the address of the storage (the "fake globals" are const i32 globals that are exported on the Wasm module, and whose value is the address of the data symbol). It wasn't expecting LLD-side changes to be necessary for that to work - is it because the fake globals aren't visible in JS for some reason, or LLD isn't making them as it should?

File: tests/test_benchmark.py [hunk 1]
```diff
@@ -183,10 +183,10 @@ def run(self, args):
 
   def get_output_files(self):
     ret = [self.filename]
-    if 'WASM=1' in self.extra_args:
-      ret.append(self.filename[:-3] + '.wasm')
-    else:
+    if 'WASM=0' in self.extra_args:
       ret.append(self.filename + '.mem')
+    else:
+      ret.append(self.filename[:-3] + '.wasm')
```
Comment (sbc100) [inline tests/test_benchmark.py]: use splitext?

File: tests/test_benchmark.py [hunk 2]
```diff
@@ -334,18 +334,17 @@ def handle_static_lib(self, f):
...
+      #EmscriptenBenchmarker('sm-asm2wasm-lto',  SPIDERMONKEY_ENGINE + ['--no-wasm-baseline'], ['--llvm-lto', '1']),
+      #EmscriptenBenchmarker('sm-wasmbackend',  SPIDERMONKEY_ENGINE + ['--no-wasm-baseline'], [env={
       #  'LLVM': '/home/alon/Dev/llvm/build/bin',
       #  'EMCC_WASM_BACKEND': '1',
       #}),
     ]
   if V8_ENGINE and Building.which(V8_ENGINE[0]):
     benchmarkers += [
-      EmscriptenBenchmarker('v8-wasm',  V8_ENGINE,           ['-s', 'WASM=1']),
+      EmscriptenBenchmarker('v8-wasm',  V8_ENGINE),
```
Comment (sbc100) [inline tests/test_benchmark.py]: Are these changes unrelated?  Perhaps mention that in the description?

File: tools/emterpretify.py [hunk 3]
```diff
@@ -61,7 +61,17 @@ def handle_arg(arg):
...
+# part of getting there - say, if an invoke was part of the path to get here, and invoke calls dynCall which
+# calls SAFE_FT_MASK - then we'll end up doing that call during recreating the stack, which breaks. In other
+# words, dynCall_* must be calls without running emterpreted code in them. To avoid that, we blacklist
+# SAFE_FT_MASK, which should be blacklisted anyhow as it has no need for emterpretation.
+BLACKLIST = set(['_malloc', '_free', '_memcpy', '_memmove', '_memset', '_strlen', 'stackAlloc', 'setThrew', 'stackRestore', 'setTempRet0', 'getTempRet0', 'stackSave', '_emscripten_autodebug_double', '_emscripten_autodebug_float', '_emscripten_autodebug_i8', '_emscripten_autodebug_i16', '_emscripten_autodebug_i32', '_emscripten_autodebug_i64', '_strncpy', '_strcpy', '_strcat', '_saveSetjmp', '_testSetjmp', '_emscripten_replace_memory', '_bitshift64Shl', '_bitshift64Ashr', '_bitshift64Lshr', 'setAsyncState', 'emtStackSave', 'emtStackRestore', 'getEmtStackMax', 'setEmtStackMax', 'SAFE_FT_MASK', 'SAFE_HEAP_LOAD', 'SAFE_HEAP_LOAD_D', 'SAFE_HEAP_STORE', 'SAFE_HEAP_STORE_D'])
+
```
Comment (sbc100) [inline tools/emterpretify.py]: Unrelated?

File: tools/ports/binaryen.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import os, shutil, logging
 
-TAG = 'version_47'
+TAG = 'version_48'
```
Comment (sbc100) [inline tools/ports/binaryen.py]: Unrelated?

**Comment by kripken** [comment]
Yes, currently JS can't get a global address from compiled code. We can add that in the future, replacing the function call here with a direct access (basically it would be an optimization, nothing else should change).

Maybe we do want to remove `makeGlobalUse()`, once we remove all uses, yeah, although perhaps we could just put a throw in it that says we intend to add that in the future.

**Comment by kripken** [comment]
All those unrelated changes (benchmark, etc.) should not have shown up. I'm not sure why they did. I think maybe I had this based on other PRs that didn't land yet? Anyhow, after merging incoming they are all gone.

**Comment by sbc100** [comment]
Thanks alon!  I would have taken me a lot longer to figure out this change myself.

**Comment by kripken** [comment]
No problem @sbc100. Should I go and do the other ones too?

**Comment by sbc100** [comment]
I'm not going to argue if you are offering :)   Thanks

**Comment by NWilson** [comment]
> @NWilson my understanding is that there is not really an existing mechanism for JS to consume global addresses directly from the compiled code (there exists some support but it looks like its focused on side modules). I imagine it could be added but it would be larger change and would have to work with existing meta-DCE system that works for functions.

Thanks, I see now. It's not that LLD doesn't support it - it's the way Emscripten uses LLD and its DCE system that needs to be able to "see" the variable being accessed from JS, and that infrastructure isn't there.

**Comment by kripken** [comment]
@sbc100 as I was starting to look at the others, I realized I don't understand something here. This should only be a problem if the symbol is defined in JS and needs to be used from C, I think? But tzname as well as the others appear to be only defined and used in JS. Where does the lld problem come in?

**Comment by sbc100** [comment]
You are correct.  However tzname is useable from C programs even outside of the standard library (via <time.h>) so it can be needed at link time.    If the global is indeed JS only then it doesn't even need to live in the heap at all right?

**Comment by kripken** [comment]
Ah right, thanks @sbc100, I forgot that crucial part...
